module/extension/Director/src/director/remove_task.cpp

```diff
@@ -36,25 +37,31 @@ namespace module::extension {
             group.active_task = nullptr;
 
             // Re-evaluate the queue of tasks for this group
-            // This may set `group.active_task` to a valid value
+            // This may set `group.active_task` to a valid value if a new task is picked up
             reevaluate_queue(group);
 
             // If nothing in the queue updated the active task to a new task we are now idle
             // That also means we need to remove any subtasks this group had recursively
             if (group.active_task == nullptr) {
                 group.state = provider::ProviderGroup::IDLE;
 
-                // TODO(@TrentHouliston)
-                //  if there is a pure leaving provider, run it
-                //  maybe I need a new set of DSL keywords so there can be a pure leaving keyword that isn't leaving
-                //  what leaving without causing/when providers do are very different from what they do with
-                //      when/causing.
-                //  maybe the leaving without a causing needs it's own DSL keyword, (might be leaving and current
-                //      leaving needs renaming)
+                // Run the Stop reactions for this provider group since it is no longer running
+                for (auto& provider : group.providers) {
+                    if (provider->classification == ProviderClassification::STOP) {
+                        auto task = provider->reaction->get_task();
+                        if (task) {
+                            task->run(std::move(task));
+                        }
```

Comment:
So getting the task and running it is how to run a provider?
Also does this imply that multiple STOP Providers might exist for one ProviderGroup?

Reply:
Getting a task and running it or pushing it to the thread pool is how you run a NUClear reaction. This particular one is me running it directly rather than handing it to the thread pool. Since the providers are NUClear reactions I'm running it here.